// _frler-scms-plugins@1.5.15@frler-scms-plugins/lib/modal.js
const modal = {
  zuiMessage: function(type, content, isAuto, cb) {
    let isAction = isAuto || isAuto == undefined ? [] : { name: 'confirm', icon: '', text: '确认', action: cb }
    new $.zui.Messager(content, {
      type: type,
      close: false,
      time: isAuto || isAuto == undefined ? 3000 : 0,
      cssClass: 'my-messagers',
      actions: isAction
    }).show()
  },

  zuiConfirm: function(confirmObj) {
    if (document.getElementById('mySmModal') == null) {
      const modalHtml = `<div class="modal fade" id="mySmModal"><div class="modal-dialog modal-sm"><div class="modal-content"><div class="modal-header"><h4 class="modal-title">确认框</h4></div><div class="modal-body"> <div style="display:table;"><div class="ib"><i class="icon icon-question-sign" style="font-size:50px;color:#F1A325;"></i></div><div style="display:table-cell;vertical-align:middle;padding:0 6px;"><span class="modalMainContent">${confirmObj.content}</span></div></div> </div><div class="modal-footer"><button type="button" class="btn btn-primary" id="confirmModalConfirmBtn">确认</button><button type="button" class="btn btn-default" data-dismiss="modal" id="confirmModalCancelBtn">取消</button></div></div></div</div>`
      $(modalHtml).appendTo($(document.body))
    } else {
      $('#mySmModal').find('.modalMainContent').text(confirmObj.content)
    }
    $('#confirmModalConfirmBtn').off('click')
    $('#confirmModalConfirmBtn').on('click', function() {
      $('#mySmModal').modal('hide')
      confirmObj.onOk()
    })
    $('#confirmModalCancelBtn').off('click')
    $('#confirmModalCancelBtn').on('click', confirmObj.onCancel)
    $('#mySmModal').modal('show', 'fit')
  },

  createIframeModal: function(options) {
    window.$modal = new $.zui.ModalTrigger({
      title: options.title,
      height: options.height || 'auto',
      width: options.width || '600px',
      data: options.data || null,
      iframe: options.iframe,
      backdrop: 'static',
      moveable: true,
      fade: true,
      successFunc: options.successFunc || null,
      cancelFunc: options.cancelFunc || this.hide
    });
    window.$modal.show({
      shown: () => {
        $('.modal-header').find('button.close').remove()
        $('.modal-content').css('height', '100%')
        $('.modal-body').css('height', 'calc(100% - 50px)')
      }
    });
  },

  closeIframeModal: function(modal) {
    modal.close()
  }
}

module.exports = modal
